panels/site_sections/attractions_admin.py

```diff
@@ -0,0 +1,330 @@
+                params,
+                bools=Attraction.all_bools,
+                checkgroups=Attraction.all_checkgroups)
+            message = check(attraction)
+            if not message:
+                if not attraction.department_id:
+                    attraction.department_id = None
+                session.add(attraction)
+                raise HTTPRedirect(
+                    'form?id={}&message={}',
+                    attraction.id,
+                    '{} updated successfully'.format(attraction.name))
+        else:
+            attraction = session.query(Attraction) \
+                .filter_by(id=attraction_id) \
+                .options(
+                    subqueryload(Attraction.department),
+                    subqueryload(Attraction.features)
+                        .subqueryload(AttractionFeature.events)
+                            .subqueryload(AttractionEvent.attendees)) \
+                .order_by(Attraction.id).one()
+
+        return {
+            'admin_account': session.current_admin_account(),
+            'message': message,
+            'attraction': attraction
+        }
+
+    def new(self, session, message='', **params):
+        if params.get('id', 'None') != 'None':
+            raise HTTPRedirect('form?id={}', params['id'])
+
+        if 'notifications' in params:
+            ns = listify(params.get('notifications', []))
+            params['notifications'] = [int(n) for n in ns if n != '']
+
+        admin_account = session.current_admin_account()
+        attraction = session.attraction(
+            params,
+            bools=Attraction.all_bools,
+            checkgroups=Attraction.all_checkgroups)
+        if not attraction.department_id:
+            attraction.department_id = None
+
+        if cherrypy.request.method == 'POST':
+            message = check(attraction)
+            if not message:
+                attraction.owner = admin_account
+                session.add(attraction)
+                raise HTTPRedirect('form?id={}', attraction.id)
+
+        return {
+            'admin_account': admin_account,
+            'attraction': attraction,
+            'message': message,
+        }
+
+    @csrf_protected
+    def delete(self, session, id, message=''):
+        if cherrypy.request.method == 'POST':
+            attraction = session.query(Attraction).get(id)
+            attendee = session.admin_attendee()
+            if not attendee.can_admin_attraction(attraction):
+                raise HTTPRedirect(
+                    'form?id={}&message={}',
+                    id,
+                    "You cannot delete an attraction that you don't own")
+
+            session.delete(attraction)
+            raise HTTPRedirect(
+                'index?message={}',
+                'The {} attraction was deleted'.format(attraction.name))
+
+        raise HTTPRedirect('form?id={}', id)
+
+    def feature(self, session, attraction_id=None, message='', **params):
```

Comment:
This is an unusual request, but I STILL have trouble keeping track of what's an attraction, what's a feature, and what's an event. At some point can you add a docstring that explains/gives an example, even if only at the top of this file?

Thinking about it, I wonder if we can/should at some point change `event` to `timeslot`. That may help a little.

Reply:
Yeah, there's lots of different types of events throughout the system. I tried to use `attraction_event` where there might be ambiguity, and just `event` where the meaning would be clear through context.

Naming is one of the hardest things in computer science!